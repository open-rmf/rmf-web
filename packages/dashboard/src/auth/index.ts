export * from './authenticator';
export * from './keycloak';
export * from './login-card';
export * from './login-page';
export * from './private-route';
export * from './stub';
export * from './user-profile';
