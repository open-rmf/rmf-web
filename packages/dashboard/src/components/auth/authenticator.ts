import Debug from 'debug';
import EventEmitter from 'eventemitter3';
import Keycloak_, { KeycloakInstance } from 'keycloak-js';
import { BASE_PATH } from '../../util/url';
import { User } from './user';

const debug = Debug('authenticator');

export type AuthenticatorEventType = {
  userChanged: [User | null];
};

export default interface Authenticator extends EventEmitter<AuthenticatorEventType> {
  readonly user?: User;
  readonly token?: string;

  /**
   * It is required to call this before using any of the authenticator functions.
   */
  init(): Promise<void>;

  /**
   * Note: This redirects to external login page so it will never return.
   */
  login(): Promise<never>;

  /**
   * Note: This redirects to external logout page so it will never return.
   */
  logout(): Promise<void>;
}

export type AuthConfig<T = typeof Keycloak_> = T extends (config: infer U) => unknown ? U : never;

export class DefaultAuthenticator
  extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  get user(): User | undefined {
    return this._user;
  }

  get token(): string | undefined {
    return this._inst.idToken;
  }

  constructor(config: AuthConfig, redirectUri?: string) {
    super();
    this._inst = Keycloak_(config);
    this._redirectUri = redirectUri;
  }

  async init() {
    debug('initializing authenticator');

    this._inst.onAuthSuccess = async () => {
      this._user = {
        username: (this._inst.idTokenParsed as any).preferred_username,
      };
      debug('authenticated as', this._user.username);
      this.emit('userChanged', this._user);
    };

    this._inst.onAuthLogout = () => {
      debug('logout');
      this._user = undefined;
      this.emit('userChanged', null);
    };

    await this._inst.init({
      onLoad: 'check-sso',
      silentCheckSsoRedirectUri: `${BASE_PATH}/silent-check-sso.html`,
      redirectUri: this._redirectUri,
    });
    try {
      const refreshed = await this._inst.updateToken(30);
      refreshed && debug('token refreshed');
    } catch {}

    this._user = this._inst.idTokenParsed && {
      username: (this._inst.idTokenParsed as any).preferred_username,
    };
  }

  async login(): Promise<never> {
    await this._inst.login();
    throw new Error('should not reach here');
  }

  async logout(): Promise<never> {
    await this._inst.logout();
    throw new Error('should not reach here');
  }

  private _inst: KeycloakInstance;
  private _redirectUri?: string;
  private _user?: User;
}
