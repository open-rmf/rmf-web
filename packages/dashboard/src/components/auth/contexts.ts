import { Permission, User as ApiUser } from 'api-client';
import React from 'react';

export type UserProfile = ApiUser;

export interface User {
  profile: UserProfile;
  permissions: Permission[];
}

export const UserContext = React.createContext<User | null>(null);
