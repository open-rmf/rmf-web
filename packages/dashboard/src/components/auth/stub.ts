import EventEmitter from 'eventemitter3';
import Authenticator, { AuthenticatorEventType } from './authenticator';
import { User } from './user';

export default class StubAuthenticator
  extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  readonly user: User = {
    username: 'stub',
  };

  readonly token?: string = undefined;

  init(): Promise<void> {
    return Promise.resolve();
  }

  login(): Promise<never> {
    throw new Error('not supported');
  }

  logout(): Promise<never> {
    throw new Error('not supported');
  }
}
