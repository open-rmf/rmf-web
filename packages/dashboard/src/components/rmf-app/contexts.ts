import { Dispenser, Fleet, Ingestor } from 'api-client';
import React from 'react';
import * as RmfModels from 'rmf-models';
import { HealthStatus } from '../../managers/rmf-health-state-manager';
import { Place } from './place';
import { RmfIngress } from './rmf-ingress';

export const BuildingMapContext = React.createContext<RmfModels.BuildingMap | null>(null);
export const PlacesContext = React.createContext<Place[]>([]);
export const FleetsContext = React.createContext<Fleet[]>([]);
export const DispensersContext = React.createContext<Dispenser[]>([]);
export const IngestorsContext = React.createContext<Ingestor[]>([]);

const itemSummary = () => {
  return {
    operational: 0,
    spoiltItem: [],
  };
};

const initializeItemSummary = itemSummary();

export const RmfHealthContext = React.createContext<HealthStatus>({
  door: { ...initializeItemSummary },
  lift: { ...initializeItemSummary },
  dispenser: { ...initializeItemSummary },
  robot: {
    operational: 0,
    idle: 0,
    charging: 0,
    spoiltRobots: [],
  },
});

export const RmfIngressContext = React.createContext<RmfIngress | undefined>(undefined);
