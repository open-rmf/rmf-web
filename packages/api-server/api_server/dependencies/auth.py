from typing import Optional

from ..authenticator import JwtAuthenticator
from ..models import User


def auth_scheme(
    authenticator: Optional[JwtAuthenticator] = None,
    oidc_url: Optional[str] = None,
):
    """
    Returns a tuple containing an authentication and user dependency. The authentication
    dependency should be applied app-wide or router-wide to verify a token. The user
    dependency should be applied on each endpoint that requires the user info.
    """
    if not authenticator:
        # no authentication
        return lambda: User(username="stub", is_admin=True)

    oidc_url = oidc_url or ""

    return authenticator.fastapi_dep(oidc_url)
