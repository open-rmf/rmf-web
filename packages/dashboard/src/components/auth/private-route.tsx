import Debug from 'debug';
import React from 'react';
import { Redirect, Route, RouteProps, useLocation } from 'react-router';
import { LOGIN_ROUTE } from '../../util/url';
import Unauthorized from '../error-pages/unauthorized';
import { UserContext } from './contexts';

const debug = Debug('PrivateRoute');

export interface PrivateRouteProps extends React.PropsWithChildren<RouteProps> {
  // if true, do not redirect to login url if not authenticated
  noRedirectToLogin?: boolean;
}

/**
 * This component validates if the user is authenticated before rendering component passed as a
 * prop.
 */
const PrivateRoute = ({
  noRedirectToLogin,
  children,
  ...rest
}: PrivateRouteProps): JSX.Element | null => {
  const user = React.useContext(UserContext);
  const location = useLocation();

  function render() {
    if (user) {
      return children;
    } else {
      if (!noRedirectToLogin) {
        debug('accessing private route while unauthenticated');
        debug('redirecting to login page');
        return <Redirect to={{ pathname: LOGIN_ROUTE, state: { from: location } }} />;
      } else {
        return <Unauthorized />;
      }
    }
  }

  return <Route {...rest}>{render()}</Route>;
};

export default PrivateRoute;
