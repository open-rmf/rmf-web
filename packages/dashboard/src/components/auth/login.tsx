import { Button } from '@material-ui/core';
import React from 'react';
import { LoginPage } from 'react-components';
import { Redirect } from 'react-router';
import { DASHBOARD_ROUTE, getUrl } from '../../util/url';
import { AuthenticatorContext, UserContext } from '../auth/contexts';

export interface LoginProps {
  /**
   * defaults to url of `DASHBOARD_ROUTE`
   */
  successRedirectUri?: string;
}

export default function Login(props: LoginProps): JSX.Element {
  const successRedirectUri = props.successRedirectUri || getUrl(DASHBOARD_ROUTE);
  const authenticator = React.useContext(AuthenticatorContext);
  const user = React.useContext(UserContext);

  async function handleRmfLogin(_event: React.MouseEvent): Promise<void> {
    authenticator.login(successRedirectUri);
  }

  return user ? (
    <Redirect to={successRedirectUri} />
  ) : (
    <LoginPage title="RoMi Dashboard" logo="assets/ros-health.png">
      <Button id="login-button" onClick={handleRmfLogin} variant="contained">
        Login with RMF
      </Button>
    </LoginPage>
  );
}
