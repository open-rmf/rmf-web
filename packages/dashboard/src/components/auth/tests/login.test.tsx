import { createMount } from '@material-ui/core/test-utils';
import React from 'react';
import { Redirect } from 'react-router';
import { BrowserRouter } from 'react-router-dom';
import { AuthenticatorContext, UserContext } from '../../auth/contexts';
import Login from '../login';
import FakeAuthenticator from '../__mocks__/authenticator';

const mount = createMount();

describe('Login page', () => {
  test('renders correctly', () => {
    const component = mount(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );
    expect(component).toMatchSnapshot();
  });

  test('redirects to dashboard when user is authenticated', () => {
    window.history.replaceState(window.history.state, '');
    const wrapper = mount(
      <BrowserRouter>
        <UserContext.Provider
          value={{
            username: 'test',
          }}
        >
          <Login />
        </UserContext.Provider>
      </BrowserRouter>,
    );
    expect(wrapper.find(Redirect)).toBeTruthy();
  });

  test('performs login when login button is clicked', () => {
    const authenticator = new FakeAuthenticator({ username: 'fakeUser' });
    const spy = jest.spyOn(authenticator, 'login').mockImplementation(() => undefined as any);

    const root = mount(
      <BrowserRouter>
        <AuthenticatorContext.Provider value={authenticator}>
          <Login />
        </AuthenticatorContext.Provider>
      </BrowserRouter>,
    );
    const loginButton = root.find('button#login-button').first();
    expect(loginButton).toBeTruthy();
    loginButton.simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
