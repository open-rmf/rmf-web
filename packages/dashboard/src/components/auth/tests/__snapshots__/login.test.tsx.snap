// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Login>
      <LoginPage
        logo="assets/ros-health.png"
        title="RoMi Dashboard"
      >
        <div
          className="makeStyles-container-1"
        >
          <ForwardRef
            logo="assets/ros-health.png"
            title="RoMi Dashboard"
          >
            <div
              className="makeStyles-container-2"
            >
              <WithStyles(ForwardRef(Typography))
                className="makeStyles-title-3"
                variant="h4"
              >
                <ForwardRef(Typography)
                  className="makeStyles-title-3"
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                  variant="h4"
                >
                  <h4
                    className="MuiTypography-root makeStyles-title-3 MuiTypography-h4"
                  >
                    RoMi Dashboard
                  </h4>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
              <img
                alt=""
                className="makeStyles-logo-4"
                src="assets/ros-health.png"
              />
              <WithStyles(ForwardRef(Button))
                id="login-button"
                onClick={[Function]}
                variant="contained"
              >
                <ForwardRef(Button)
                  classes={
                    Object {
                      "colorInherit": "MuiButton-colorInherit",
                      "contained": "MuiButton-contained",
                      "containedPrimary": "MuiButton-containedPrimary",
                      "containedSecondary": "MuiButton-containedSecondary",
                      "containedSizeLarge": "MuiButton-containedSizeLarge",
                      "containedSizeSmall": "MuiButton-containedSizeSmall",
                      "disableElevation": "MuiButton-disableElevation",
                      "disabled": "Mui-disabled",
                      "endIcon": "MuiButton-endIcon",
                      "focusVisible": "Mui-focusVisible",
                      "fullWidth": "MuiButton-fullWidth",
                      "iconSizeLarge": "MuiButton-iconSizeLarge",
                      "iconSizeMedium": "MuiButton-iconSizeMedium",
                      "iconSizeSmall": "MuiButton-iconSizeSmall",
                      "label": "MuiButton-label",
                      "outlined": "MuiButton-outlined",
                      "outlinedPrimary": "MuiButton-outlinedPrimary",
                      "outlinedSecondary": "MuiButton-outlinedSecondary",
                      "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                      "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                      "root": "MuiButton-root",
                      "sizeLarge": "MuiButton-sizeLarge",
                      "sizeSmall": "MuiButton-sizeSmall",
                      "startIcon": "MuiButton-startIcon",
                      "text": "MuiButton-text",
                      "textPrimary": "MuiButton-textPrimary",
                      "textSecondary": "MuiButton-textSecondary",
                      "textSizeLarge": "MuiButton-textSizeLarge",
                      "textSizeSmall": "MuiButton-textSizeSmall",
                    }
                  }
                  id="login-button"
                  onClick={[Function]}
                  variant="contained"
                >
                  <WithStyles(ForwardRef(ButtonBase))
                    className="MuiButton-root MuiButton-contained"
                    component="button"
                    disabled={false}
                    focusRipple={true}
                    focusVisibleClassName="Mui-focusVisible"
                    id="login-button"
                    onClick={[Function]}
                    type="button"
                  >
                    <ForwardRef(ButtonBase)
                      className="MuiButton-root MuiButton-contained"
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "focusVisible": "Mui-focusVisible",
                          "root": "MuiButtonBase-root",
                        }
                      }
                      component="button"
                      disabled={false}
                      focusRipple={true}
                      focusVisibleClassName="Mui-focusVisible"
                      id="login-button"
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="MuiButtonBase-root MuiButton-root MuiButton-contained"
                        disabled={false}
                        id="login-button"
                        onBlur={[Function]}
                        onClick={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <span
                          className="MuiButton-label"
                        >
                          Login with RMF
                        </span>
                        <WithStyles(memo)
                          center={false}
                        >
                          <ForwardRef(TouchRipple)
                            center={false}
                            classes={
                              Object {
                                "child": "MuiTouchRipple-child",
                                "childLeaving": "MuiTouchRipple-childLeaving",
                                "childPulsate": "MuiTouchRipple-childPulsate",
                                "ripple": "MuiTouchRipple-ripple",
                                "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                "rippleVisible": "MuiTouchRipple-rippleVisible",
                                "root": "MuiTouchRipple-root",
                              }
                            }
                          >
                            <span
                              className="MuiTouchRipple-root"
                            >
                              <TransitionGroup
                                childFactory={[Function]}
                                component={null}
                                exit={true}
                              />
                            </span>
                          </ForwardRef(TouchRipple)>
                        </WithStyles(memo)>
                      </button>
                    </ForwardRef(ButtonBase)>
                  </WithStyles(ForwardRef(ButtonBase))>
                </ForwardRef(Button)>
              </WithStyles(ForwardRef(Button))>
            </div>
          </ForwardRef>
        </div>
      </LoginPage>
    </Login>
  </Router>
</BrowserRouter>
`;
